### PR TITLE
Add gradle-plugin init type

### PR DESCRIPTION
**Pull request by big-guy**
@eriwen here's what I was thinking.  The sample project isn't very complicated (custom plugin, custom task), but we could expand it (to include an extension or whatever we needed to show).

This PR adds documentation (release notes + user guide), a new init type (gradle-plugin) and an integration test.  The templates are derived from the automaticClasspathInjectionCustomTestSourceSet TestKit sample.

### CODE REVIEW

File: subprojects/build-init/src/main/groovy/org/gradle/buildinit/plugins/internal/GradlePluginProjectInitDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.gradle.buildinit.plugins.internal.BuildInitTestFramework.SPOCK;
+
+public class GradlePluginProjectInitDescriptor extends LanguageLibraryProjectInitDescriptor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GradlePluginProjectInitDescriptor.class);
+
+    public GradlePluginProjectInitDescriptor(TemplateOperationFactory templateOperationFactory,
+                                            FileResolver fileResolver,
```
Comment (eriwen) [inline subprojects/build-init/src/main/groovy/org/gradle/buildinit/plugins/internal/GradlePluginProjectInitDescriptor.java]: Nitpick: I've never seen white-spacing that lined variables with the left parenthesis. Usually they align to beginning of first type declaration.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/FunctionalTest.groovy.template [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import spock.lang.Specification
+
+import static org.gradle.testkit.runner.TaskOutcome.*
+
+/**
+ * Functional tests for the Hello World plugin.
+ */
+class FunctionalTest extends Specification {
```
Comment (eriwen) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/FunctionalTest.groovy.template]: Nitpick: seems like we'd want to name this class `HelloWorldFunctionalTest`

**Comment by eriwen** [comment]
This is a great start. I think it'd be much more useful if we expand it as you say. 

Additions that I think would help ensure users' success:
 * An extension + tests
 * An incremental task
 * Use of [IncrementalTaskInputs](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html)
 * A simple integration test
 * Use of `gradle/` folder to separate parts of Gradle config, for example to break out functional and integration tests. (I love how @bmuschko has done this in his plugins)
 * Plugin publishing

I think we should establish as many good conventions to users through this mechanism as we can, which is why I suggest so many additions.

Like I said, this is a great start. Other than the optional nitpicks, we can merge this as far as I'm concerned. You're welcome to get a 2nd opinion of course :)

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/build.gradle.template [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// In this section you declare where to find the dependencies of your project
+repositories {
+    // Use 'jcenter' for resolving your dependencies.
+    // You can declare any Maven/Ivy/file repository here.
+    jcenter()
+}
+
+// Functional tests using Gradle TestKit.
+// Read more about this in the user guide ${ref_userguide_testkit.groovyComment}
+sourceSets {
```
Comment (bmuschko) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/build.gradle.template]: IMHO setting up the functional test source set + conventions should be provided by the Java Gradle plugin. We already declare the TestKit dependency automatically. Why not also setting up appropriate source set conventions?

**Comment by wolfs** [comment]
@eriwen How do you want to proceed with this PR? Are you gonna merge that soon?

**Comment by eriwen** [comment]
@wolfs I have been in contact with @big-guy and we will come up with the best way to move forward with this in the next couple days. IMO I think we want to do some work on this before we push and announce it.

**Comment by bmuschko** [comment]
@eriwen Did you guys have a chance to talk in the meantime?

**Comment by oehme** [comment]
To make sense of an extension, we need to go beyond hello world. An extension only makes sense if there is something to abstract above the tasks. If we just have a hello task and a hello extension with the same fields, users will be thoroughly confused why we make it so complicated.

My suggestion would be a database migration example. The extension defines databases and the plugin creates a drop task and a migrate task for every database. The tasks themselves would of course just do some println.

IncrementalTaskInputs is a nice to have imho. Just having inputs and outputs declared at all would probably be a good start for most people.

**Comment by eriwen** [comment]
I think @oehme is right that better guides would be preferable for now. Having one `init` template to rule them all is not the way to go because we can't predict our users' needs. 

I'd like to close this and link it to our to-be-announced projects. Any objections?

**Comment by oehme** [comment]
I'm not saying not to do an init template, but to keep it very simple. A plugin class that adds a task and a test that exercises it, plus the build.gradle file to build and publish it on the portal. No extensions or other advanced stuff. Also no separation of unit and functional tests if TestKit doesn't do it out of the box. I think such very basic scaffolding would be useful. The rest can be done with guides.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/build.gradle.template [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+check.dependsOn functionalTest
+
+gradlePlugin {
+    testSourceSets sourceSets.functionalTest
+    plugins {
+        // Declare your plugins with id and implementation.
+        helloWorld {
+            id = 'helloworld'
```
Comment (oehme) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/gradleplugin/build.gradle.template]: This needs to be globally unique, so let's do something like org.example.hello

**Comment by pledbrook** [comment]
Following on from @oehme's comments, project templates are not tutorials or samples, so shouldn't be treated as such.

**Comment by bmuschko** [comment]
Based on the discussions I am going to close the PR. Most of the code generated by the proposed PR could probably covered by getting started guides. Those guides would provide far more context and would be less confusing to the end user. If we decide to enhance the `init` functionality, it should be as slimed down version.

**Pull request closed without merging**